<a href="https://colab.research.google.com/github/jonmessier/Sales-Predictions/blob/main/Project_1_Part2(_Core).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Intro
Project 1 - Part 2 (core)\
Jon Messier \
11/16/2022 \

[Coding Dojo Learn](https://login.codingdojo.com/m/606/13762/97764)

We will continue to work off of the GitHub repository you created last week for your first portfolio project. The goal of your portfolio projects are to have items to show potential employers that demonstrate your data science skills and competencies.

The first project will be a sales prediction for food items sold at various stores. The goal of this is to help the retailer understand the properties of products and outlets that play crucial roles in increasing sales.

For Part 2, your task is to use Pandas to load in the data and start exploring and cleaning it.  As data scientists, about 80% of our time is spent on data preparation - so this is a very important step!

Download the data using this link: [download the data](https://drive.google.com/file/d/1syH81TVrbBsdymLT_jl2JIf6IjPXtSQw/view?usp=sharing).

(Note: [Original data source](https://datahack.analyticsvidhya.com/contest/practice-problem-big-mart-sales-iii/)). \
 \



#Data Dictionary 
<table><tbody><tr><td><strong>Variable Name</strong></td><td><strong>Description</strong></td></tr><tr><td>Item_Identifier</td><td>Unique product ID</td></tr><tr><td>Item_Weight</td><td>Weight of product</td></tr><tr><td>Item_Fat_Content</td><td>Whether the product is low fat or regular</td></tr><tr><td>Item_Visibility</td><td>The percentage of total display area of all products in a store allocated to the particular product</td></tr><tr><td>Item_Type</td><td>The category to which the product belongs</td></tr><tr><td>Item_MRP</td><td>Maximum Retail Price (list price) of the product</td></tr><tr><td>Outlet_Identifier</td><td>Unique store ID</td></tr><tr><td>Outlet_Establishment_Year</td><td>The year in which store was established</td></tr><tr><td>Outlet_Size</td><td>The size of the store in terms of ground area covered</td></tr><tr><td>Outlet_Location_Type</td><td>The type of area in which the store is located</td></tr><tr><td>Outlet_Type</td><td>Whether the outlet is a grocery store or some sort of supermarket</td></tr><tr><td>Item_Outlet_Sales</td><td>Sales of the product in the particular store. This is the target variable to be predicted.
<a href="https://github.com/ShauryaBhandari/Bigmart-Sales-Prediction#why-does-the-the-problem-need-to-be-solved" id="user-content-why-does-the-the-problem-need-to-be-solved" class="anchor" aria-hidden="true" target="_blank"></a></td></tr></tbody></table>

# Imports
* Use Pandas to read the sales prediction data set into Google Colab.
* Use df.info() and df.head() methods to preview the first rows of your data and a summary of your DataFrame's columns.



In [38]:
import pandas as pd
import numpy as np

In [39]:
df = pd.read_csv('/content/sales_predictions.csv')
print(df.head())
print(df.info())

  Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier

## 1) How many rows and columns?

In [40]:
print('(Rows, Columns)')
df.shape

(Rows, Columns)


(8523, 12)

## 2) What are the datatypes of each variable?

In [41]:
print("The columns of the the sales prediction file have the following data types")
display(df.dtypes)


The columns of the the sales prediction file have the following data types


Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

#Data Cleaning
  * Explore the data - what do you need to do to clean this data? Clean it!
  * Be sure to address each of the following tasks:


## 3) Are there duplicates? If so, drop any duplicates.

In [42]:
print(f'Number of duplicates: {df.duplicated().sum()}')

Number of duplicates: 0


In [43]:
#If we had duplicates
#df.drop_duplicates(inplace=True)

## 4) Identify missing values
From the df.isna().sum() we can see that there a couple of columns with Null data
- 'Item_Weight' has 1463 null entries
- 'Outlet_Size' has 2410 null entries

In [44]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

## 5) Address Missing Values
Decide on how to address the missing values and do it! (This requires your judgement, so explain your choice).

### 5.A 'Item_Weight'
Begin with a statitistical look at the Item_Weight. This is a numerical column so using a calculated approach to fill the missing data is appropriate.  

In [45]:
# 5 a) Begin with the statistics on 'Item_Weight'. 
df.groupby('Item_Type')['Item_Weight'].describe().round(2)

count   mean   std   min    25%    50%    75%    max
Item_Type                                                                   
Baking Goods            536.0  12.28  4.77  4.88   8.24  11.65  15.75  20.85
Breads                  204.0  11.35  4.44  4.64   7.13  10.60  14.85  20.85
Breakfast                89.0  12.77  5.04  6.42   8.06  10.70  17.25  21.10
Canned                  539.0  12.31  4.59  4.62   8.12  12.15  16.25  21.35
Dairy                   566.0  13.43  4.69  4.80   9.28  13.35  17.60  20.70
Frozen Foods            718.0  12.87  4.51  4.56   8.94  12.85  17.00  20.85
Fruits and Vegetables  1019.0  13.22  4.58  5.46   9.20  13.10  17.10  21.35
Hard Drinks             183.0  11.40  4.24  4.61   8.26  10.10  14.85  19.70
Health and Hygiene      430.0  13.14  4.51  5.18   9.70  12.15  17.60  21.25
Household               759.0  13.38  5.00  5.03   8.70  13.15  18.35  21.25
Meat                    337.0  12.82  4.67  5.15   9.30  12.35  17.20  21.25
Others                  137.0  13.85  4.23  5.50  10.65  14.50  17.75  20.50
Seafood                  51.0  12.55  5.47  5.36   7.42  11.65  17.75  20.75
Snack Foods             988.0  12.99  4.53  5.10   9.20  13.15  16.85  21.25
Soft Drinks             374.0  11.85  4.40  4.59   8.26  11.80  15.35  20.75
Starchy Foods           130.0  13.69  4.01  6.70  11.50  13.18  16.70  21.20

>After reviewing the statistics on the Item Weights by Item_Type, I can see that the values have a normal distribution about the mean with min() and max() within 2std.  This looks like a close sample of data, there adding to the mean will have little impact.  Replace missing Item_Weights with the mean() value of the Item_Type weight.  


In [46]:
#Replace the NaN weights with the mean
df['Item_Weight'].fillna(df['Item_Weight'].mean(),inplace=True)
#display the results compared to previous output.  Notice that the mean doesnt change much, but the count goes up to include non-Nan values
df.groupby('Item_Type')['Item_Weight'].describe().round(2)

count   mean   std   min    25%    50%    75%    max
Item_Type                                                                   
Baking Goods            648.0  12.38  4.35  4.88   9.10  12.86  15.15  20.85
Breads                  251.0  11.63  4.04  4.64   8.00  12.80  14.08  20.85
Breakfast               110.0  12.79  4.53  6.42   8.27  12.86  16.50  21.10
Canned                  649.0  12.40  4.18  4.62   8.63  12.86  15.70  21.35
Dairy                   682.0  13.33  4.27  4.80  10.80  12.86  16.74  20.70
Frozen Foods            856.0  12.87  4.13  4.56   9.30  12.86  16.10  20.85
Fruits and Vegetables  1232.0  13.16  4.16  5.46  10.17  12.86  16.35  21.35
Hard Drinks             214.0  11.61  3.95  4.61   8.37  11.23  14.65  19.70
Health and Hygiene      520.0  13.09  4.10  5.18  10.10  12.86  16.60  21.25
Household               910.0  13.30  4.57  5.03   9.10  12.86  17.21  21.25
Meat                    425.0  12.83  4.16  5.15  10.20  12.86  15.35  21.25
Others                  169.0  13.66  3.82  5.50  11.15  12.86  17.35  20.50
Seafood                  64.0  12.61  4.88  5.36   7.97  12.73  17.35  20.75
Snack Foods            1200.0  12.96  4.11  5.10   9.60  12.86  16.00  21.25
Soft Drinks             445.0  12.01  4.05  4.59   8.52  12.86  15.20  20.75
Starchy Foods           148.0  13.59  3.77  6.70  12.15  12.86  16.18  21.20

>Comparing the above output to the dataframe containing NaN values you can see a a large increase in the item count and and a small increase mean 'Item_Weight' for all item types.  This makes sense since we are adding many entries where the Item_Weight is mean value of the NaN data set. 

### 5.B 'Outlet_Size'
'Outlet_Size' is a categorical column, so we must impart our wisdom on to the missing values.  Let's look at the list of values and their counts to determine how we fill the empty values

In [47]:
print(df['Outlet_Size'].value_counts())
print('Number of NaN Outlet_Size values: ')
df['Outlet_Size'].isna().sum()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64
Number of NaN Outlet_Size values: 


2410

>There are 2410 missing Outlet size entries.  The distribution of Outlet Size are Medium (53%), small(46%), High(11%). Since we're not really doing anything analytic with this data, and we do not know any context behind the values we're missing we cannot setup a reasonable model for inserting values.  Therefore we will fill the missing values with "Unkown"


In [49]:
df['Outlet_Size'].fillna('Unknown', inplace=True)
df['Outlet_Size'].value_counts()

Medium     2793
Unknown    2410
Small      2388
High        932
Name: Outlet_Size, dtype: int64

### 6) Confirm that there are no missing values after addressing them.

In [50]:
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

## 7) Inconsistent Categories
Find and fix any inconsistent categories of data (example: fix cat, Cat, and cats so that they are consistent).
- Categorical data is a non-numeric data type : 'object'
- Look at the datatypes of all columns.  If there is a column that you think is *numeric* but shows as an *object*, you could try converting it to a numeric type.  this would create an error(s) on non-numeric entries.  You could go back in and update those entries manually.

In [73]:
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

Create a list of all columns with *object* data type

In [54]:
data_types = df.dtypes
cols = data_types[data_types=='object'].index
cols

Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

Loop through all the columns of data that have an object data type and output the count unique values to see if there are some changes we can/should make.  Use the *df['col].value_counts()* function.  This lets me see if data was entered correctly.

In [60]:
for name in cols:
 print(f"The unique values of the feature {name} are:\n")
 print(df[name].value_counts())
 print("\n")

The unique values of the feature Item_Identifier are:

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64


The unique values of the feature Item_Fat_Content are:

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64


The unique values of the feature Item_Type are:

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64


The unique values of the 

Looks like there is some cleanup needed in the "Item_Fat_Content" feature.  We can change the outliers:
- **'LF (316)' and 'low fat (112)' --> 'Low Fat (5089)'**
- **'reg'(117) --> 'Regular'(2889)** 

Use *df.replace()* to update these mistyped entries.  Show the updated column unique value counts

In [61]:
df.replace(to_replace=['LF', 'low fat'], value='Low Fat', inplace=True)
df.replace('reg','Regular', inplace=True)
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

## 8) Statistics
For any numerical columns, obtain the summary statistics of each (min, max, mean). We only want to look at numeric datatypes

In [ ]:
types = df.dtypes

<bound method NDFrame.describe of Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object>

In [62]:
df.select_dtypes(include=np.number).describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
count      8523.00          8523.00   8523.00                    8523.00   
mean         12.86             0.07    140.99                    1997.83   
std           4.23             0.05     62.28                       8.37   
min           4.56             0.00     31.29                    1985.00   
25%           9.31             0.03     93.83                    1987.00   
50%          12.86             0.05    143.01                    1999.00   
75%          16.00             0.09    185.64                    2004.00   
max          21.35             0.33    266.89                    2009.00   

       Item_Outlet_Sales  
count            8523.00  
mean             2181.29  
std              1706.50  
min                33.29  
25%               834.25  
50%              1794.33  
75%              3101.30  
max             13086.96

Reviewing the statitistics, we do not have an large outliers in the numerical columns.  No changes to the numeric data needed at this time